src/Microsoft.Restier.Core/ApiBuilderExtensions.cs

```diff
@@ -186,48 +183,91 @@ public static ApiBuilder ChainPrevious<T>(this ApiBuilder obj, Func<IServiceProv
-        /// Constructor parameters other than the next service instance would be fulfilled from service provider.
-        /// And this method is capable of choosing a constructor base on available services.
+        /// The contributor added will get an instance of <typeparamref name="TImplement"/> from the container, i.e.
+        /// <see cref="IServiceProvider"/>, every time it's get called.
+        /// This method will try to register <typeparamref name="TImplement"/> as a service with
+        /// <see cref="ServiceLifetime.Transient"/> life time, if it's not yet registered. To override, you can
+        /// register <typeparamref name="TImplement"/> before or after calling this method.
         /// </summary>
+        /// <remarks>
+        /// Note: When registering <typeparamref name="TImplement"/>, you must NOT give it a
+        /// <see cref="ServiceLifetime"/> that makes it outlives <typeparamref name="TService"/>, that could possibly
+        /// make an instance of <typeparamref name="TImplement"/> be used in multiple instantiations of
+        /// <typeparamref name="TService"/>, which leads to unpredictable behaviors.
+        /// </remarks>
         /// <typeparam name="TService">The service type.</typeparam>
         /// <typeparam name="TImplement">The implementation type.</typeparam>
         /// <param name="obj">The <see cref="ApiBuilder"/>.</param>
         /// <returns>Current <see cref="ApiBuilder"/></returns>
         public static ApiBuilder ChainPrevious<TService, TImplement>(this ApiBuilder obj)
             where TService : class
-            where TImplement : TService
+            where TImplement : class, TService
         {
-            Func<IServiceProvider, TService, TService> factory = null;
+            Ensure.NotNull(obj, "obj");
+
+            Func<IServiceProvider, Func<TService>, TService> factory = null;
 
-            return obj.AddContributor<TService>((sp, next) =>
+            obj.Services.TryAddTransient<TImplement>();
+            return obj.AddContributorNoCheck<TService>((sp, next) =>
             {
                 if (factory != null)
                 {
-                    return factory(sp, next());
+                    return factory(sp, next);
                 }
 
-                // TODO: See if in the future ActivatorUtilities could replace these implementation. For now its
-                // methods can't select constructor appropriately, base on avaialble services.
-                var ret = ChainedService.CreateViaReflection(typeof(TService), typeof(TImplement), sp, next);
-                var value = ret.Item1;
-                var creator = ret.Item2;
+                var instance = sp.GetService<TImplement>();
+                if (instance == null)
+                {
+                    return instance;
+                }
+
+                var nextProperty = typeof(TImplement).GetTypeInfo().GetProperties()
```

Comment:
**Thank you!**
I don't even know it's possible. I supposed that expression compilation needs the property to be public settable. I was an idiot...